src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2499,10 +2499,13 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
             max_block_size = std::max<UInt64>(1, max_block_limited);
             max_threads_execute_query = max_streams = 1;
         }
+
         if (local_limits.local_limits.size_limits.max_rows != 0)
         {
             if (max_block_limited < local_limits.local_limits.size_limits.max_rows)
                 query_info.limit = max_block_limited;
+            else if (local_limits.local_limits.size_limits.max_rows < std::numeric_limits<UInt64>::max()) /// Ask to read just enough rows to make the max_rows limit effective (so it has a chance to be triggered).
```

Comment:
In new analyzer the same logic is here
https://github.com/ClickHouse/ClickHouse/blob/bf2edb972e897a9993e2a047c9666514c595876c/src/Planner/PlannerJoinTree.cpp#L689-L694

Setting `table_expression_query_info.limit` there fixes `03024_total_rows_approx_is_set_for_system_zeros_and_generate_random`